src/iceberg/schema.h

```diff
@@ -54,13 +56,42 @@ class ICEBERG_EXPORT Schema : public StructType {
 
   [[nodiscard]] std::string ToString() const override;
 
+  [[nodiscard]] std::optional<std::reference_wrapper<const SchemaField>> GetFieldByName(
+      std::string_view name, bool case_sensitive) const override;
+
+  [[nodiscard]] std::optional<std::reference_wrapper<const SchemaField>> GetFieldByName(
+      std::string_view name) const;
+
+  [[nodiscard]] std::optional<std::reference_wrapper<const SchemaField>> GetFieldById(
+      int32_t field_id) const override;
+
   friend bool operator==(const Schema& lhs, const Schema& rhs) { return lhs.Equals(rhs); }
 
+  mutable std::unordered_map<int, size_t> id_to_index_;
+  mutable std::unordered_map<std::string, size_t> name_to_index_;
+  mutable std::unordered_map<std::string, size_t> lowercase_name_to_index_;
+  mutable std::vector<SchemaField> full_schemafield_;
+
  private:
   /// \brief Compare two schemas for equality.
   [[nodiscard]] bool Equals(const Schema& other) const;
 
   const std::optional<int32_t> schema_id_;
+
+  void InitIdToIndexMap() const;
+  void InitNameToIndexMap() const;
+  void InitLowerCaseNameToIndexMap() const;
 };
 
+class SchemaFieldVisitor {
```

Comment:
Can we declare it in the schema.cc only? This class is not supposed to be used by users.